README.rst
=========
aioapp_pg
=========

Micro framework based on asyncio

.. image:: https://img.shields.io/pypi/v/aioapp_pg.svg
        :target: https://pypi.python.org/pypi/aioapp_pg

.. image:: https://img.shields.io/travis/inplat/aioapp_pg.svg
        :target: https://travis-ci.org/inplat/aioapp_pg

.. image:: https://codecov.io/gh/inplat/aioapp_pg/branch/master/graph/badge.svg
        :target: https://codecov.io/gh/inplat/aioapp_pg

.. image:: https://pyup.io/repos/github/inplat/aioapp_pg/shield.svg
        :target: https://pyup.io/repos/github/inplat/aioapp_pg/
        :alt: Updates

* Free software: Apache License 2.0





HISTORY.rst
=======
History
=======

0.0.1b1 (2018-01-17)
--------------------

* First release on PyPI.


tests/test_db.py
import asyncio
from aioapp.app import Application
from aioapp_pg import Postgres, PostgresTracerConfig
from aioapp.error import PrepareError
import pytest
import string
from aioapp.misc import rndstr
from aioapp.tracer import Span
from aioapp.misc import async_call


async def _start_postgres(app: Application, url: str,
                          connect_max_attempts=10,
                          connect_retry_delay=1.0) -> Postgres:
    db = Postgres(url, connect_max_attempts=connect_max_attempts,
                  connect_retry_delay=connect_retry_delay)
    app.add('db', db)
    await app.run_prepare()
    await db.start()
    return db


def _create_span(app) -> Span:
    if app.tracer:
        return app.tracer.new_trace(sampled=False, debug=False)


async def test_postgres(app, postgres):
    table_name = 'tbl_' + rndstr(20, string.ascii_lowercase + string.digits)
    db = await _start_postgres(app, postgres)
    span = _create_span(app)

    res = await db.execute(span, 'test',
                           'SELECT $1::int as a', 1, timeout=10,
                           tracer_config=PostgresTracerConfig())
    assert res is not None

    res = await db.query_one(span, 'test',
                             'SELECT $1::int as a, $2::json, $3::jsonb', 1, {},
                             {}, timeout=10)
    assert res is not None
    assert len(res) == 3
    assert res[0] == 1
    assert res[1] == {}
    assert res[2] == {}
    assert res['a'] == 1

    res = await db.query_one(span, 'test',
                             'SELECT $1::int as a WHERE FALSE', 1, timeout=10)
    assert res is None

    res = await db.query_all(span, 'test',
                             'SELECT UNNEST(ARRAY[$1::int, $2::int]) as a',
                             1, 2, timeout=10)
    assert res is not None
    assert len(res) == 2
    assert res[0][0] == 1
    assert res[1][0] == 2
    assert res[0]['a'] == 1
    assert res[1]['a'] == 2

    res = await db.query_all(span, 'test',
                             'SELECT $1::int as a WHERE FALSE', 1, timeout=10)
    assert res is not None
    assert len(res) == 0

    async with db.connection(span) as conn:
        async with conn.xact(span):
            await conn.execute(span, 'test',
                               'CREATE TABLE %s(id int);' % table_name)
            await conn.execute(span, 'test',
                               'INSERT INTO %s(id) VALUES(1)' % table_name)

    res = await db.query_one(span, 'test',
                             'SELECT COUNT(*) FROM %s' % table_name,
                             timeout=10)
    assert res[0] == 1

    try:
        async with db.connection(span) as conn:
            async with conn.xact(span, isolation_level='SERIALIZABLE'):
                await conn.execute(span, 'test',
                                   'INSERT INTO %s(id) VALUES(2)' % table_name)
                raise UserWarning()
    except UserWarning:
        pass

    res = await db.query_one(span, 'test',
                             'SELECT COUNT(*) FROM %s' % table_name,
                             timeout=10)
    assert res[0] == 1


async def test_postgres_prepare_failure(app, unused_tcp_port):
    with pytest.raises(PrepareError):
        await _start_postgres(app, 'postgres://postgres@%s:%s/postgres'
                                   '' % ('127.0.0.1', unused_tcp_port),
                              connect_max_attempts=2,
                              connect_retry_delay=0.001)


async def test_postgres_health_bad(app: Application, unused_tcp_port: int,
                                   loop: asyncio.AbstractEventLoop) -> None:
    url = 'postgres://postgres@%s:%s/postgres' % ('127.0.0.1', unused_tcp_port)

    db = Postgres(url)
    app.add('postgres', db)

    async def start():
        await app.run_prepare()
        await db.start()

    res = async_call(loop, start)
    await asyncio.sleep(1)

    result = await app.health()
    assert 'postgres' in result
    assert result['postgres'] is not None
    assert isinstance(result['postgres'], BaseException)

    if res['fut'] is not None:
        res['fut'].cancel()


async def test_postgres_health_ok(app: Application, postgres: str,
                                  loop: asyncio.AbstractEventLoop) -> None:
    db = Postgres(postgres)
    app.add('postgres', db)

    async def start():
        await app.run_prepare()
        await db.start()

    res = async_call(loop, start)
    await asyncio.sleep(1)

    result = await app.health()
    assert 'postgres' in result
    assert result['postgres'] is None

    if res['fut'] is not None:
        res['fut'].cancel()


tests/__init__.py
# -*- coding: utf-8 -*-

"""Unit test package for aioapp."""


tests/conftest.py
import gc
import time
import asyncio
import socket
import pytest
import asyncpg
from compose.service import ImageType
from compose.project import Project
from aioapp.app import Application

COMPOSE_POSTGRES_PORT = 19811


@pytest.fixture(scope='session')
def event_loop():
    asyncio.set_event_loop_policy(asyncio.DefaultEventLoopPolicy())
    loop = asyncio.get_event_loop_policy().new_event_loop()
    yield loop
    gc.collect()
    loop.close()


@pytest.fixture(scope='session')
def loop(event_loop):
    return event_loop


def pytest_addoption(parser):
    parser.addoption("--postgres-addr", dest="postgres_addr",
                     help="Use this postgres instead of docker image "
                          "if specified",
                     metavar="postgres://user:passwd@host:port/dbname")
    parser.addoption('--show-docker-logs', dest="show_docker_logs",
                     action='store_true', default=False,
                     help='Show docker logs after test')


@pytest.fixture(scope='session')
def postgres_override_addr(request):
    return request.config.getoption('postgres_addr')


@pytest.fixture(scope='session')
async def docker_compose(loop, request,
                         docker_project: Project,
                         postgres_override_addr):
    async def check_postgres(url):
        conn = await asyncpg.connect(url, loop=loop)
        await conn.close()

    checks = {
        (
            'postgres',
            'POSTGRES_DSN',
            postgres_override_addr,
            'postgresql://postgres@127.0.0.1:%d/postgres'
            '' % COMPOSE_POSTGRES_PORT,
            check_postgres
        ),
    }

    result = {}

    fns = []
    to_start = []
    for svc, name, override, url, fn in checks:
        if override:
            result[name] = override
        else:
            to_start.append(svc)
            fns.append((fn, url))
            result[name] = url

    if not to_start:
        yield result
    else:
        containers = docker_project.up(to_start)

        if not containers:
            raise ValueError("`docker-compose` didn't launch any containers!")

        try:
            timeout = 60
            start_time = time.time()
            print()
            print('Waiting for docker services...')
            last_err = None
            while start_time + timeout > time.time():
                try:
                    await asyncio.gather(*[fn(url) for fn, url in fns],
                                         loop=loop)
                    break

                except Exception as err:
                    last_err = err
                    await asyncio.sleep(1, loop=loop)
            else:
                last_err_type = type(last_err)
                raise TimeoutError(f'Unable to start all container services'
                                   f' within {timeout} seconds. Last error:'
                                   f' {last_err} ({last_err_type})')
            print('Docker services are ready')
            yield result
        finally:

            # Send container logs to stdout, so that they get included in
            # the test report.
            # https://docs.pytest.org/en/latest/capture.html
            for container in sorted(containers, key=lambda c: c.name):
                if request.config.getoption('show_docker_logs'):
                    header = f"Logs from {container.name}:"
                    print(header)
                    print("=" * len(header))
                    print(
                        container.logs().decode("utf-8", errors="replace") or
                        "(no logs)"
                    )
                    print()

            docker_project.down(ImageType.none, False)


@pytest.fixture(scope='session')
def postgres(docker_compose):
    return docker_compose['POSTGRES_DSN']


def get_free_port(protocol='tcp'):
    family = socket.AF_INET
    if protocol == 'tcp':
        type = socket.SOCK_STREAM
    elif protocol == 'udp':
        type = socket.SOCK_DGRAM
    else:
        raise UserWarning()

    sock = socket.socket(family, type)
    try:
        sock.bind(('', 0))
        return sock.getsockname()[1]
    finally:
        sock.close()


@pytest.fixture()
async def app(loop):
    app = Application(loop=loop)
    yield app
    await app.run_shutdown()


examples/pg.py
import os
import time
import logging
import asyncio
from functools import partial
from aioapp.app import Application
from aioapp import config
from aioapp.tracer import Span
from aioapp_pg import Postgres


class Config(config.Config):
    db_url: str
    _vars = {
        'db_url': {
            'type': str,
            'name': 'DB_URL',
            'descr': 'Database connection string in following format '
                     'postgresql://user:passwd@host:port/dbname'
        }
    }


async def do_something(app: Application, ctx: Span) -> None:
    """
    do not run this task infinitely!!!
    there is no graceful shutdown!
    """
    res = await app.db.query_one(
        ctx, 'example_query_id',
        'SELECT $1::float as now, pg_sleep(1)',
        time.time())
    print('query result', res['now'])


if __name__ == '__main__':
    logging.basicConfig(level=logging.INFO)
    loop = asyncio.get_event_loop()

    cfg = Config(os.environ)

    app = Application(loop=loop)
    app.add(
        'db',
        Postgres(
            url=cfg.db_url,
            pool_min_size=2,
            pool_max_size=19,
            pool_max_queries=50000,
            pool_max_inactive_connection_lifetime=300.,
            connect_max_attempts=10,
            connect_retry_delay=1.0),
        stop_after=[]
    )
    app.on_start = partial(do_something, app)
    app.run()


travis_pypi_setup.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
"""Update encrypted deploy password in Travis config file."""


from __future__ import print_function
import base64
import json
import os
from getpass import getpass
import yaml
from cryptography.hazmat.primitives.serialization import load_pem_public_key
from cryptography.hazmat.backends import default_backend
from cryptography.hazmat.primitives.asymmetric.padding import PKCS1v15


try:
    from urllib import urlopen
except ImportError:
    from urllib.request import urlopen


GITHUB_REPO = 'inplat/aioapp_pg'
TRAVIS_CONFIG_FILE = os.path.join(
    os.path.dirname(os.path.abspath(__file__)), '.travis.yml')


def load_key(pubkey):
    """Load public RSA key.

    Work around keys with incorrect header/footer format.

    Read more about RSA encryption with cryptography:
    https://cryptography.io/latest/hazmat/primitives/asymmetric/rsa/
    """
    try:
        return load_pem_public_key(pubkey.encode(), default_backend())
    except ValueError:
        # workaround for https://github.com/travis-ci/travis-api/issues/196
        pubkey = pubkey.replace('BEGIN RSA', 'BEGIN').replace('END RSA', 'END')
        return load_pem_public_key(pubkey.encode(), default_backend())


def encrypt(pubkey, password):
    """Encrypt password using given RSA public key and encode it with base64.

    The encrypted password can only be decrypted by someone with the
    private key (in this case, only Travis).
    """
    key = load_key(pubkey)
    encrypted_password = key.encrypt(password, PKCS1v15())
    return base64.b64encode(encrypted_password)


def fetch_public_key(repo):
    """Download RSA public key Travis will use for this repo.

    Travis API docs: http://docs.travis-ci.com/api/#repository-keys
    """
    keyurl = 'https://api.travis-ci.org/repos/{0}/key'.format(repo)
    data = json.loads(urlopen(keyurl).read().decode())
    if 'key' not in data:
        errmsg = "Could not find public key for repo: {}.\n".format(repo)
        errmsg += "Have you already added your GitHub repo to Travis?"
        raise ValueError(errmsg)
    return data['key']


def prepend_line(filepath, line):
    """Rewrite a file adding a line to its beginning."""
    with open(filepath) as f:
        lines = f.readlines()

    lines.insert(0, line)

    with open(filepath, 'w') as f:
        f.writelines(lines)


def load_yaml_config(filepath):
    """Load yaml config file at the given path."""
    with open(filepath) as f:
        return yaml.load(f)


def save_yaml_config(filepath, config):
    """Save yaml config file at the given path."""
    with open(filepath, 'w') as f:
        yaml.dump(config, f, default_flow_style=False)


def update_travis_deploy_password(encrypted_password):
    """Put `encrypted_password` into the deploy section of .travis.yml."""
    config = load_yaml_config(TRAVIS_CONFIG_FILE)

    config['deploy']['password'] = dict(secure=encrypted_password)

    save_yaml_config(TRAVIS_CONFIG_FILE, config)

    line = ('# This file was autogenerated and will overwrite'
            ' each time you run travis_pypi_setup.py\n')
    prepend_line(TRAVIS_CONFIG_FILE, line)


def main(args):
    """Add a PyPI password to .travis.yml so that Travis can deploy to PyPI.

    Fetch the Travis public key for the repo, and encrypt the PyPI password
    with it before adding, so that only Travis can decrypt and use the PyPI
    password.
    """
    public_key = fetch_public_key(args.repo)
    password = args.password or getpass('PyPI password: ')
    update_travis_deploy_password(encrypt(public_key, password.encode()))
    print("Wrote encrypted password to .travis.yml -- you're ready to deploy")


if '__main__' == __name__:
    import argparse
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument('--repo', default=GITHUB_REPO,
                        help='GitHub repo (default: %s)' % GITHUB_REPO)
    parser.add_argument('--password',
                        help='PyPI password (will prompt if not provided)')

    args = parser.parse_args()
    main(args)


aioapp_pg/__init__.py
import json
import time
import traceback
from typing import Union, Dict, List, Any, Optional
import asyncio
import asyncpg
import asyncpg.protocol
import asyncpg.pool
from aioapp.app import Component
from aioapp.error import PrepareError
from aioapp.misc import mask_url_pwd
from aioapp.tracer import (Span, CLIENT, SPAN_TYPE, SPAN_KIND)

SPAN_TYPE_POSTGRES = 'postgres'
SPAN_KIND_POSTRGES_ACQUIRE = 'acquire'
SPAN_KIND_POSTRGES_QUERY = 'query'

__version__ = '0.0.1a1'

JsonType = Union[None, int, float, str, bool, List[Any], Dict[str, Any]]


class PostgresTracerConfig:

    def on_acquire_start(self, ctx: 'Span') -> None:
        pass

    def on_acquire_end(self, ctx: 'Span',
                       err: Optional[Exception]) -> None:
        if err:
            ctx.tag('error.message', str(err))
            ctx.annotate(traceback.format_exc())

    def on_query_start(self, ctx: 'Span', id: str, query: str,
                       args: tuple, timeout: Optional[float]) -> None:
        pass

    def on_query_end(self, ctx: 'Span',
                     err: Optional[Exception], result) -> None:
        if err:
            ctx.tag('error.message', str(err))
            ctx.annotate(traceback.format_exc())


class Postgres(Component):
    def __init__(self, url: str, pool_min_size: int = 10,
                 pool_max_size: int = 10,
                 pool_max_queries: int = 50000,
                 pool_max_inactive_connection_lifetime: float = 300.0,
                 connect_max_attempts: int = 10,
                 connect_retry_delay: float = 1.0) -> None:
        super(Postgres, self).__init__()
        self.url = url
        self.pool_min_size = pool_min_size
        self.pool_max_size = pool_max_size
        self.pool_max_queries = pool_max_queries
        self.pool_max_inactive_connection_lifetime = \
            pool_max_inactive_connection_lifetime
        self.connect_max_attempts = connect_max_attempts
        self.connect_retry_delay = connect_retry_delay
        self._pool: asyncpg.pool.Pool = None
        self._connections: List['Connection'] = []

    @property
    def pool(self) -> asyncpg.pool.Pool:
        return self._pool

    @property
    def _masked_url(self) -> Optional[str]:
        if self.url is not None:
            return mask_url_pwd(self.url)

    async def _connect(self) -> None:
        if self.app is None:
            raise UserWarning('Unattached component')

        self.app.log_info("Connecting to %s" % self._masked_url)
        self._pool: asyncpg.pool.Pool = await asyncpg.create_pool(
            dsn=self.url,
            max_size=self.pool_max_size,
            min_size=self.pool_min_size,
            max_queries=self.pool_max_queries,
            max_inactive_connection_lifetime=(
                self.pool_max_inactive_connection_lifetime),
            init=Postgres._conn_init,
            loop=self.loop
        )
        self.app.log_info("Connected to %s" % self._masked_url)

    @staticmethod
    async def _conn_init(conn: asyncpg.pool.PoolConnectionProxy) -> None:
        def _json_encoder(value: JsonType) -> str:
            return json.dumps(value)

        def _json_decoder(value: str) -> JsonType:
            return json.loads(value)

        await conn.set_type_codec(
            'json', encoder=_json_encoder, decoder=_json_decoder,
            schema='pg_catalog'
        )

        def _jsonb_encoder(value: JsonType) -> bytes:
            return b'\x01' + json.dumps(value).encode('utf-8')

        def _jsonb_decoder(value: bytes) -> JsonType:
            return json.loads(value[1:].decode('utf-8'))

        # Example was got from https://github.com/MagicStack/asyncpg/issues/140
        await conn.set_type_codec(
            'jsonb',
            encoder=_jsonb_encoder,
            decoder=_jsonb_decoder,
            schema='pg_catalog',
            format='binary',
        )

    async def prepare(self) -> None:
        if self.app is None:
            raise UserWarning('Unattached component')

        for i in range(self.connect_max_attempts):
            try:
                await self._connect()
                return
            except Exception as e:
                self.app.log_err(str(e))
                await asyncio.sleep(self.connect_retry_delay)
        raise PrepareError("Could not connect to %s" % self._masked_url)

    async def start(self) -> None:
        pass

    async def stop(self) -> None:
        if self.app is None:
            raise UserWarning('Unattached component')

        stop_timeout = 60
        stop_start = time.time()

        xact_locks = [conn._xact_lock.acquire() for conn in self._connections
                      if conn._xact_lock is not None]
        if len(xact_locks) > 0:
            await asyncio.wait(xact_locks, loop=self.loop,
                               timeout=max(
                                   stop_timeout - stop_start + stop_start,
                                   0.001))

        conn_locks = [conn._lock.acquire() for conn in self._connections
                      if conn._lock is not None]
        if len(conn_locks):
            await asyncio.wait(conn_locks, loop=self.loop,
                               timeout=max(
                                   stop_timeout - stop_start + stop_start,
                                   0.001))

        if self.pool:
            self.app.log_info("Disconnecting from %s" % self._masked_url)
            await self.pool.close()

    def connection(self, ctx: Span,
                   acquire_timeout=None,
                   tracer_config: Optional[PostgresTracerConfig] = None
                   ) -> 'ConnectionContextManager':
        return ConnectionContextManager(self, ctx,
                                        acquire_timeout=acquire_timeout,
                                        tracer_config=tracer_config)

    async def query_one(self, ctx: Span, id: str, query: str,
                        *args: Any, timeout: float = None,
                        tracer_config: Optional[PostgresTracerConfig] = None
                        ) -> asyncpg.protocol.Record:
        async with self.connection(ctx,
                                   tracer_config=tracer_config) as conn:
            return await conn.query_one(ctx, id, query, *args,
                                        timeout=timeout,
                                        tracer_config=tracer_config)

    async def query_all(self, ctx: Span, id: str, query: str,
                        *args: Any, timeout: float = None,
                        tracer_config: Optional[PostgresTracerConfig] = None
                        ) -> List[asyncpg.protocol.Record]:
        async with self.connection(ctx,
                                   tracer_config=tracer_config) as conn:
            return await conn.query_all(ctx, id, query, *args,
                                        timeout=timeout,
                                        tracer_config=tracer_config)

    async def execute(self, ctx: Span, id: str, query: str,
                      *args: Any, timeout: float = None,
                      tracer_config: Optional[PostgresTracerConfig] = None
                      ) -> str:
        async with self.connection(ctx,
                                   tracer_config=tracer_config) as conn:
            return await conn.execute(ctx, id, query, *args,
                                      timeout=timeout,
                                      tracer_config=tracer_config)

    async def health(self, ctx: Span):
        async with self.connection(ctx) as conn:
            await conn.execute(ctx, 'test', 'SELECT 1')


class ConnectionContextManager:
    def __init__(self, db: Postgres, ctx: Span,
                 acquire_timeout: float = None,
                 tracer_config: Optional[PostgresTracerConfig] = None) -> None:
        self._db = db
        self._conn = None
        self._ctx = ctx
        self._acquire_timeout = acquire_timeout
        self._tracer_config = tracer_config
        self._pg_conn: Optional['Connection'] = None

    async def __aenter__(self) -> 'Connection':
        span = None
        if self._ctx:
            span = self._ctx.new_child()
        try:
            if span:
                span.kind(CLIENT)
                span.name("db:Acquire")
                span.metrics_tag(SPAN_TYPE, SPAN_TYPE_POSTGRES)
                span.metrics_tag(SPAN_KIND, SPAN_KIND_POSTRGES_ACQUIRE)
                span.remote_endpoint("postgres")
                span.start()
                if self._tracer_config:
                    self._tracer_config.on_acquire_start(span)
            self._conn = await self._db._pool.acquire(
                timeout=self._acquire_timeout)
            if span:
                if self._tracer_config:
                    self._tracer_config.on_acquire_end(span, None)
                span.finish()
        except Exception as err:
            if span:
                if self._tracer_config:
                    self._tracer_config.on_acquire_end(span, err)
                span.finish(exception=err)
            raise
        self._pg_conn = Connection(self._db, self._conn)
        self._db._connections.append(self._pg_conn)
        return self._pg_conn

    async def __aexit__(self, exc_type: type, exc: BaseException,
                        tb: type) -> bool:
        if self._pg_conn is not None:
            self._db._connections.remove(self._pg_conn)
        await self._db._pool.release(self._conn)
        return False


class TransactionContextManager:
    def __init__(self, ctx: Span, conn: 'Connection',
                 isolation_level: str = None,
                 xact_lock: asyncio.Lock = None,
                 tracer_config: Optional[PostgresTracerConfig] = None) -> None:
        self._conn = conn
        self._isolation_level = isolation_level
        self._ctx = ctx
        self._tracer_config = tracer_config
        self._xact_lock = xact_lock
        self._in_transaction = False

    def _begin_query(self) -> str:
        query = "BEGIN TRANSACTION"
        if self._isolation_level:
            query += " ISOLATION LEVEL %s" % self._isolation_level
        return query

    async def __aenter__(self) -> None:
        if self._in_transaction:
            raise UserWarning('Transaction already started')
        if self._xact_lock is not None:
            await self._xact_lock.acquire()
        self._in_transaction = True
        await self._conn.execute(self._ctx,
                                 query=self._begin_query(),
                                 id="BeginTransaction",
                                 tracer_config=self._tracer_config)

    async def __aexit__(self, exc_type: type, exc: BaseException,
                        tb: type) -> bool:
        try:
            if exc:
                await self._conn.execute(self._ctx,
                                         query="ROLLBACK", id="Rollback",
                                         tracer_config=self._tracer_config)
            else:
                await self._conn.execute(self._ctx,
                                         query="COMMIT", id="Commit",
                                         tracer_config=self._tracer_config)
        finally:
            self._in_transaction = False
            if self._xact_lock is not None:
                self._xact_lock.release()
        return False


class Connection:
    def __init__(self, db: Postgres,
                 conn: asyncpg.pool.PoolConnectionProxy) -> None:
        self._db = db
        self._conn = conn
        self._lock = asyncio.Lock(loop=db.loop)
        self._xact_lock = asyncio.Lock(loop=db.loop)

    def xact(self, ctx: Span,
             isolation_level: str = None,
             tracer_config: Optional[PostgresTracerConfig] = None
             ) -> 'TransactionContextManager':
        return TransactionContextManager(ctx, self, isolation_level,
                                         self._xact_lock,
                                         tracer_config)

    async def execute(self, ctx: Span, id: str,
                      query: str, *args: Any, timeout: float = None,
                      tracer_config: Optional[
                          PostgresTracerConfig] = None) -> str:
        with await self._lock:
            span = None
            if ctx:
                span = ctx.new_child()

            try:
                if span:
                    span.kind(CLIENT)
                    span.name("db:%s" % id)
                    span.metrics_tag(SPAN_TYPE, SPAN_TYPE_POSTGRES)
                    span.metrics_tag(SPAN_KIND, SPAN_KIND_POSTRGES_QUERY)
                    span.remote_endpoint("postgres")
                    span.annotate(repr(args))
                    span.start()
                    if tracer_config:
                        tracer_config.on_query_start(span, id, query, args,
                                                     timeout)
                res = await self._conn.execute(query, *args, timeout=timeout)
                if span:
                    if tracer_config:
                        tracer_config.on_query_end(span, None, res)
                    span.finish()
            except Exception as err:
                if span:
                    if tracer_config:
                        tracer_config.on_query_end(span, err, None)
                    span.finish(exception=err)
                raise

            return res

    async def query_one(self, ctx: Span, id: str,
                        query: str, *args: Any,
                        timeout: float = None,
                        tracer_config: Optional[PostgresTracerConfig] = None
                        ) -> asyncpg.protocol.Record:
        with await self._lock:
            span = None
            if ctx:
                span = ctx.new_child()
            try:
                if span:
                    span.kind(CLIENT)
                    span.name("db:%s" % id)
                    span.metrics_tag(SPAN_TYPE, SPAN_TYPE_POSTGRES)
                    span.metrics_tag(SPAN_KIND, SPAN_KIND_POSTRGES_QUERY)
                    span.remote_endpoint("postgres")
                    span.annotate(repr(args))
                    span.start()
                    if tracer_config:
                        tracer_config.on_query_start(span, id, query, args,
                                                     timeout)
                res = await self._conn.fetchrow(query, *args, timeout=timeout)
                if span:
                    if tracer_config:
                        tracer_config.on_query_end(span, None, res)
                    span.finish()
            except Exception as err:
                if span:
                    if tracer_config:
                        tracer_config.on_query_end(span, err, None)
                    span.finish(exception=err)
                raise
            return res

    async def query_all(self, ctx: Span, id: str,
                        query: str, *args: Any, timeout: float = None,
                        tracer_config: Optional[PostgresTracerConfig] = None
                        ) -> List[asyncpg.protocol.Record]:
        with await self._lock:
            span = None
            if ctx:
                span = ctx.new_child()
            try:
                if span:
                    span.kind(CLIENT)
                    span.name("db:%s" % id)
                    span.metrics_tag(SPAN_TYPE, SPAN_TYPE_POSTGRES)
                    span.metrics_tag(SPAN_KIND, SPAN_KIND_POSTRGES_QUERY)
                    span.remote_endpoint("postgres")
                    span.annotate(repr(args))
                    span.start()
                    if tracer_config:
                        tracer_config.on_query_start(span, id, query, args,
                                                     timeout)
                res = await self._conn.fetch(query, *args, timeout=timeout)
                if span:
                    if tracer_config:
                        tracer_config.on_query_end(span, None, res)
                    span.finish()
            except Exception as err:
                if span:
                    if tracer_config:
                        tracer_config.on_query_end(span, err, None)
                    span.finish(exception=err)
                raise
            return res
